Python/Module6_Testing/Hypothesis_Practice_Exercises.md

```diff
@@ -0,0 +1,231 @@
+>>> test_broken_identity()
+Falsifying example: test_broken_identity(
+    x=nan,
+)
+```
+
+that "NaN" (which stands for [Not a Number](https://en.wikipedia.org/wiki/NaN)) is not equal to itself.
+This is, in fact, [the designed behavior of NaN](https://en.wikipedia.org/wiki/NaN#Comparison_with_NaN) in the specification for floating point numbers.
+
+Now let's updated our strategy for describing floating point numbers to exclude this.
+
+```python
+@given(x=st.floats(allow_nan=False))
+def test_fixed_identity(x):
+    assert x == x
+```
+
+Assuming that NaN is the only floating point that violates the self-identity, this test should now pass.
+
+```python
+>>> test_fixed_identity()
+```
+<!-- #endregion -->
+
+<!-- #region -->
+**Exercise: Using Hypothesis to learn about floats.. Part 2**
+
+```python
+# using `st.floats` to find value(s) that violate `x < x + 1`
+
+@given(x=st.floats(min_value=0))
+def test_broken_inequality(x):
+    assert x < x + 1
+```
+```python
+>>> test_broken_inequality()
+Falsifying example: test_broken_inequality(
+    x=9007199254740992.0,
+)
+```
+We can check that:
+
+```python
+>>> import math
+>>> math.log2(9007199254740992)
+53
+```
+
+Recall that `2 ** 53` is the maximum size that the coefficient of a floating point double can take on. Thus Hypothesis is pointing out that `2 ** 53 + 1` can't be represented as a 64-bit floating point number.
+Let's see what happens when we do try to add one to `2 ** 53`:
+
+```python
+>>> x = 2.0 ** 53; x
+9007199254740992.0
+
+>>> x + 1
+9007199254740992.0
+```
+
+The value doesn't change at all because we would have to exceed 64-bits allotted to represent a floating point number.
+Thus `x < x + 1` fails to hold for `x = 2.0 ** 53`. 
+
+
+Now let's update our test to test everything up to this maximum value.
+Note that we want to ensure that we are testing the maximum permitted value every time we run the test as hypothesis does not guarantee that it will do so.
+We leverage the `example` decorator to accomplish this.
+
+```python
+# updating our usage of `st.floats` to generate only values that satisfy `x < x + 1`
+from hypothesis import example
+
+
+@example(x=2.0 ** 53 - 1)  # ensures that maximum permissible value is tested
+@given(x=st.floats(min_value=0, max_value=2.0 ** 53, exclude_max=True))
+def test_fixed_inequality(x):
+    assert x < x + 1
+```
+
+```python
+# running our test
+>>> test_fixed_inequality()
+```
+<!-- #endregion -->
+
+<!-- #region -->
+**Stitching together strategies for rich behavior**
+
+```python
+squares = st.integers().map(lambda x: x ** 2)
+cubes = st.integers().map(lambda x: x ** 3)
+
+# Recall that the pipe operator, `|`, is a convenient way for 
+# calling `st.one_of(...)` on strategies
+squares_or_cubes = st.tuples(squares, squares) | st.tuples(cubes, cubes, cubes)
+```
+<!-- #endregion -->
+
+<!-- #region -->
+**Describing objects that evaluate to `False`**
+
+```python
+falsies = st.sampled_from([False, 0, 0.0, "", [], tuple(), {}])
```

Comment:
```suggestion
falsies = st.sampled_from([False, 0, 0.0, "", [], (), {}])
```

I'd also add a comment explicitly noting that the same objects will be generated every time, so this strategy is unsafe if you mutate them.  Instead, you could rely on the fact that calling the respective type objects with no arguments returns the empty/falsy value: `st.sampled_from([bool, int, float, str, list, tuple, dict]).map(lambda type_: type_())`.